### PR TITLE
Checkout: Hide contact validation errors during auto-completion, with external state

### LINKED ISSUE #69211: [Bug]: "We could not validate your contact information" error on checkout load when buying a domain for a user who previously bought a non-domain
### Quick summary

If a user who already paid for a non-domain goes to buy a domain later, they get an error message like this immediately when the checkout screen loads (before they have typed anything in):

validation-error-on-checkout-load

### Steps to reproduce

The easiest way to reproduce this is to sign up with a new user account and buy a monthly plan (since that does not come with a free domain), then later go to Upgrades -> Domains and add a domain to your shopping cart.

### What you expected to happen

No error message immediately when the checkout page loads.

### What actually happened

I saw an error message immediately after the checkout page loaded.

### Browser

_No response_

### Context

_No response_

### Platform (Simple, Atomic, or both?)

_No response_

### Other notes

_No response_

### Reproducibility

Consistent

### Severity

_No response_

### Available workarounds?

_No response_

### Workaround details

_No response_

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Proposed Changes

When checkout loads, it fetches saved contact details from an API endpoint, pre-fills the contact form with those details, and attempts to auto-complete the contact details step. Since https://github.com/Automattic/wp-calypso/pull/64344, this process uses the same logic as if a user filled out the form and clicked "Continue" manually. 

However, there is a downside of this approach: the contact details are validated every time someone clicks "Continue" and since this uses the same process, if the saved details are invalid the user will see an error reported as soon as checkout loads. Since the user didn't actually take any action at this point, this can be very confusing.

validation-error

In this PR, we refactor the placement of the auto-completion code so that its validation call will not report any errors on failure. This is done by moving the auto-completion into the parent component of the `CheckoutStepGroup` where the validation function is run and refusing to show a validation error message if auto-completion is still in-progress.

Fixes https://github.com/Automattic/wp-calypso/issues/69211

#### Differences compared to the previous version of this PR

This is the second attempt at solving this issue after https://github.com/Automattic/wp-calypso/pull/69943 was reverted in https://github.com/Automattic/wp-calypso/pull/70721 because there was a bug with free purchases. 

In the previous PR, we moved the checkout contact details step into a wrapper component due to a leaky abstraction. There is one case where no contact details step is needed: when the purchase is free ($0 total, not full credits). Before https://github.com/Automattic/wp-calypso/pull/69943, we did not render the step in this case, but after that PR, we _did_ render the step, except that the wrapper returned null. The result was that, if the purchase was free, checkout would think that the contact details step was present and _active_ even though it was null, effectively blocking checkout.

This PR performs additional refactoring to remove the leaky abstraction and gets rid of the wrapper. In addition, https://github.com/Automattic/wp-calypso/pull/70776 adds a test to prevent the bug from happening again.

Some of the changes originally made for this PR have been merged separately with https://github.com/Automattic/wp-calypso/pull/70776 (which improves the tests) and https://github.com/Automattic/wp-calypso/pull/70778 (which refactors checkout steps to externalize the store).

#### Testing Instructions

There are unit tests for this behavior but you can test it manually as follows:

First test that successful validation still works:

- Make sure you have valid cached _domain_ contact details for a user (using a domain product will cause more fields to be visible). You can do this by visiting checkout when a domain product is in your cart, filling out the contact information step and pressing "Continue". Assuming they are valid, they will be saved for future visits to checkout.
- Add a _domain_ product to your cart and visit checkout (or reload if you are already there).
- Verify that you see the contact information step load before it is auto-completed.
- Verify that the active step is now the payment method step.

Next test that manual validation displays error messages:

- Add a _domain_ product to your cart and visit checkout.
- Click to edit the contact information step if it is not already active.
- Enter invalid contact information (eg: by clearing the "Last name" field).
- Press "Continue".
- Verify that you see a validation error.
- Verify that the active step is still the contact information step.

Finally, test that auto-completion does not display error messages:

- Apply D92139-code and sandbox the API. This will break the domain contact details validation endpoint.
- Add a _domain_ product to your cart and visit checkout.
- Verify that you see the contact information step load before it is auto-completed.
- Verify that you _do not see_ a validation error.
- Verify that the active step is still the contact information step.

### CODE REVIEW

**Comment by spraveenitpro** [comment]
<details>

<summary>Running unit tests</summary>



###  client/my-sites/checkout/composite-checkout/test/checkout-contact-step.js

Tests passed with a warning:



### client/my-sites/checkout/composite-checkout/test/checkout-contact-step.js

**Tests passed with a warning:**



</details>

<details>

<summary>Running functional tests</summary>

<details>

<summary>First test that successful validation still works and verify that the active step is now the payment method step:</summary>

**Works as described.**


</details>
<details>
<summary>Manual validation displays error messages. Verify that the active step is still the contact information step</summary>

**Works as described.**


</details>
<details>
<summary>Auto-completion does not display error messages. Verify that the active step is still the contact information step.:</summary>

**Works as described.**


</details>
</details>

File: client/my-sites/checkout/composite-checkout/components/wp-checkout.tsx [hunk 1]
```diff
@@ -237,6 +239,13 @@ export default function WPCheckout( {
 	const [ is3PDAccountConsentAccepted, setIs3PDAccountConsentAccepted ] = useState( false );
 	const [ isSubmitted, setIsSubmitted ] = useState( false );
 
+	const stepGroupStore = useMemo( () => createCheckoutStepGroupStore(), [] );
```
Comment (pottedmeat) [inline client/my-sites/checkout/composite-checkout/components/wp-checkout.tsx:248]: The empty `deps` threw me off a little. (No change expected).

**Review by pottedmeat — APPROVED** [review]
Changes makes sense, code looks good, type checking I did all looks good.

**Comment by pottedmeat** [comment]
Running into a problem with the unit test. `CheckoutStepGroupInner` is calling `setTotalSteps` which calls `onStateChange` which runs the callback in `store.subscription.subscribe` which calls `setContextValue` which triggers a redraw in the current execution loop.

If I make the call to `onStateChange` async (e.g. `setTimeout( onStateChange, 0 );`) the test passes, though I don't know what side-effects that might create.

**Comment by sirbrillig** [comment]
> Running into a problem with the unit test. CheckoutStepGroupInner is calling setTotalSteps which calls onStateChange which runs the callback in store.subscription.subscribe which calls setContextValue which triggers a redraw in the current execution loop.

Good point.

> If I make the call to onStateChange async (e.g. setTimeout( onStateChange, 0 );) the test passes, though I don't know what side-effects that might create.

Interesting idea. I tried it in two places, `onStateChange` and inside `notifySubscribers` (which seems maybe more appropriate?) but in both cases the unit tests fail (`yarn test-packages composite-checkout`). This may just be an issue where the unit tests are expecting the UI to update immediately and instead it takes an additional render, but my hope was that this change would modify as little behavior as possible.

I wonder if it's possible to just not call `onStateChange` inside `setTotalSteps`? Originally, `setTotalSteps` happened in an async call itself, but it wasn't fast enough, so I made it run on every render and short-circuit if there are no changes. Maybe there's no longer any need to re-render if the total number of steps changes since that's more of an internal bookkeeping thing. Trying this in https://github.com/Automattic/wp-calypso/pull/70737/commits/9cf6af93ac01564bbfec799e5fbe51b5910a7182

**Comment by sirbrillig** [comment]
Nope, that broke the auto-complete. I have another idea, to defer the call to `setState` itself. That seems to work and also resolve the warnings that @spraveenitpro mentioned.

**Comment by jjchrisdiehl** [comment]
Ran through the test steps and after loading the cart with a domain subscription, the 'Enter your contact information' step auo-completes and loads without error and the cart awaits me to press Continue. Upon pressing Continue, the Last Name field fails validation (as expected due to D92139-code).

Unit tests also pass without any warnings.

**Comment by sirbrillig** [comment]
I was seeing an unexpected delay when checkout loaded while the contact details were fetched; during the delay, the contact form was editable and active, and had no evidence that a fetch was taking place until it suddenly autocompleted the form. I think this delay actually exists in production too, but may only show up when the contact details endpoint is very slow.

To fix it, I did two things:

1. I refactored `useCachedContactDetails` to always wait until fetching was complete before returning its data.
2. I changed checkout to display its loading page until the cached contact details fetching is complete (along with all the other things it waits for).

**Comment by sirbrillig** [comment]
Oh, but now the original bug has returned. This is quite a roller coaster. 😓 Back to the drawing board.

Part of the problem is that to prevent the bug we need to wait until the steps have loaded, but if we use the checkout group store to do this (see https://github.com/Automattic/wp-calypso/pull/70737/commits/ca807bb0248e5e5cb905d43c8a9f419c44eddea0), then we are trapped because the store (by design!) lives outside of React and so when it changes nothing is re-rendered.

**Comment by sirbrillig** [comment]
Closing in favor of https://github.com/Automattic/wp-calypso/pull/71024

**Pull request closed without merging**